### PR TITLE
fix(ui,api): let users add/edit/delete workflow icon when as code

### LINKED ISSUE #4157: fix(ui,api): let users add/edit/delete workflow icon when as code

### PR DESCRIPTION
**Pull request by bnjjj**
1. Description
+ Refactor of advanced view in workflow page
1. Related issues
fix #4157 
1. About tests
UT
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/api/workflow.go [hunk 1]
```diff
@@ -490,6 +492,70 @@ func (api *API) putWorkflowHandler() service.Handler {
...
+// putWorkflowIconHandler updates a workflow
+func (api *API) putWorkflowIconHandler() service.Handler {
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+		vars := mux.Vars(r)
+		key := vars["key"]
+		name := vars["permWorkflowName"]
+
+		p, errP := project.Load(api.mustDB(), api.Cache, key, deprecatedGetUser(ctx))
+		if errP != nil {
+			return sdk.WrapError(errP, "putWorkflowHandler> Cannot load Project %s", key)
```
Comment (sguiheux) [inline engine/api/workflow.go]: return errP is enough

File: engine/api/workflow.go [hunk 2]
```diff
@@ -490,6 +492,70 @@ func (api *API) putWorkflowHandler() service.Handler {
...
+		if !strings.HasPrefix(icon, sdk.IconFormat) {
+			return sdk.ErrIconBadFormat
+		}
+		if len(icon) > sdk.MaxIconSize {
+			return sdk.ErrIconBadSize
+		}
+
+		wf, err := workflow.Load(ctx, api.mustDB(), api.Cache, p, name, deprecatedGetUser(ctx), workflow.LoadOptions{})
+		if err != nil {
+			return sdk.WrapError(err, "cannot load workflow %s", name)
```
Comment (sguiheux) [inline engine/api/workflow.go]: return err

File: engine/api/workflow.go [hunk 3]
```diff
@@ -490,6 +492,70 @@ func (api *API) putWorkflowHandler() service.Handler {
...
+		}
+
+		if err := workflow.UpdateIcon(api.mustDB(), wf.ID, icon); err != nil {
+			return sdk.WrapError(err, "cannot update workflow icon for %s", wf.Name)
```
Comment (sguiheux) [inline engine/api/workflow.go]: return err

File: engine/api/workflow.go [hunk 4]
```diff
@@ -490,6 +492,70 @@ func (api *API) putWorkflowHandler() service.Handler {
...
+// deleteWorkflowIconHandler updates a workflow
+func (api *API) deleteWorkflowIconHandler() service.Handler {
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+		vars := mux.Vars(r)
+		key := vars["key"]
+		name := vars["permWorkflowName"]
+
+		p, errP := project.Load(api.mustDB(), api.Cache, key, deprecatedGetUser(ctx))
+		if errP != nil {
+			return sdk.WrapError(errP, "putWorkflowHandler> Cannot load Project %s", key)
```
Comment (sguiheux) [inline engine/api/workflow.go]: return errP

File: engine/api/workflow.go [hunk 5]
```diff
@@ -490,6 +492,70 @@ func (api *API) putWorkflowHandler() service.Handler {
...
+		}
+
+		wf, err := workflow.Load(ctx, api.mustDB(), api.Cache, p, name, deprecatedGetUser(ctx), workflow.LoadOptions{})
+		if err != nil {
+			return sdk.WrapError(err, "cannot load workflow %s", name)
```
Comment (sguiheux) [inline engine/api/workflow.go]: return err

File: engine/api/workflow.go [hunk 6]
```diff
@@ -490,6 +492,70 @@ func (api *API) putWorkflowHandler() service.Handler {
...
+		}
+
+		if err := workflow.UpdateIcon(api.mustDB(), wf.ID, ""); err != nil {
+			return sdk.WrapError(err, "cannot update workflow icon for %s", wf.Name)
```
Comment (sguiheux) [inline engine/api/workflow.go]: return err

File: engine/api/workflow/dao.go [hunk 7]
```diff
@@ -100,6 +100,15 @@ func CountVariableInWorkflow(db gorp.SqlExecutor, projectKey string, varName str
 	return datas, nil
 }
 
+// UpdateIcon update the icon of a workflow
+func UpdateIcon(db gorp.SqlExecutor, workflowID int64, icon string) error {
+	if _, err := db.Exec("update workflow set icon = $1 where id = $2", icon, workflowID); err != nil {
+		return err
```
Comment (richardlt) [inline engine/api/workflow/dao.go]: wrap the error

**Pull request merged**